File: src/main/java/dev/lampirg/consultationappointment/service/notification/NotificationAspect.java
package dev.lampirg.consultationappointment.service.notification;

import dev.lampirg.consultationappointment.data.student.Student;
import dev.lampirg.consultationappointment.data.teacher.DatePeriod;
import dev.lampirg.consultationappointment.data.teacher.Teacher;
import dev.lampirg.consultationappointment.data.teacher.TeacherRepository;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.After;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.springframework.context.annotation.Profile;
import org.springframework.mail.SimpleMailMessage;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Aspect
@Component
@Profile("email")
@Async
public class NotificationAspect {

    private EmailService<SimpleMailMessage> emailService;
    private TeacherRepository teacherRepository;

    public NotificationAspect(EmailService<SimpleMailMessage> emailService, TeacherRepository teacherRepository) {
        this.emailService = emailService;
        this.teacherRepository = teacherRepository;
    }

    @After("execution(* dev.lampirg.consultationappointment.service.student.AppointmentMaker.makeAppointment(..)) && " +
            "args(teacher,..,datePeriod)")
    @Transactional(readOnly = true)
    public void notifyTeacherAboutAppointment(Teacher teacher, DatePeriod datePeriod) {
        if (!datePeriod.getAppointments().isEmpty())
            return;
        SimpleMailMessage message = new SimpleMailMessage();
        message.setFrom("consultation_service@pgups.example");
        message.setSubject("ПГУПС: запись на консультацию");
        message.setTo(teacher.getEmail());
        String text = """
                Здравствуйте, %s %s %s.
                На вашу консультацию, которая состоится %td.%tm с %tR до %tR записался студент.
                """;
        message.setText(String.format(text,
                teacher.getLastName(), teacher.getFirstName(), teacher.getPatronymic(),
                datePeriod.getStartTime().toLocalDate(), datePeriod.getStartTime().toLocalDate(),
                datePeriod.getStartTime().toLocalTime(), datePeriod.getEndTime().toLocalTime()
        ));
        emailService.sendEmail(message);
    }

    @After("execution(* dev.lampirg.consultationappointment.service.teacher.ConsultationMaker.deleteConsultation(..))")
    @Transactional(readOnly = true)
    public void notifyStudentsAboutDeletion(JoinPoint joinPoint) throws Throwable {
        DatePeriod datePeriod = (DatePeriod) joinPoint.getArgs()[1];
        SimpleMailMessage template = new SimpleMailMessage();
        template.setFrom("consultation_service@pgups.example");
        template.setSubject("ПГУПС: запись на консультацию");
        String text = """
                Здравствуйте, %s %s %s.
                К сожалению, консультация на которую вы записались (преподаватель: %s %s %s, дата: %td.%tm, время: с %tR до %tR), была отменена.
                """;


        datePeriod.getAppointments().forEach((appointment -> {
            Student student = appointment.getStudent();
            Teacher teacher = appointment.getTeacher();
            template.setTo(student.getEmail());
            template.setText(String.format(text,
                    student.getLastName(), student.getFirstName(), student.getPatronymic(),
                    teacher.getLastName(), teacher.getFirstName(), teacher.getPatronymic(),
                    datePeriod.getStartTime().toLocalDate(), datePeriod.getStartTime().toLocalDate(),
                    datePeriod.getStartTime().toLocalTime(), datePeriod.getEndTime().toLocalTime()
            ));
            emailService.sendEmail(template);
        }));
    }

}

File: src/main/kotlin/dev/lampirg/consultationappointment/service/notification/NotificationAspect.kt
package dev.lampirg.consultationappointment.service.notification

import dev.lampirg.consultationappointment.data.appointment.Appointment
import dev.lampirg.consultationappointment.data.teacher.DatePeriod
import dev.lampirg.consultationappointment.data.teacher.Teacher
import org.aspectj.lang.JoinPoint
import org.aspectj.lang.annotation.After
import org.aspectj.lang.annotation.Aspect
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Profile
import org.springframework.mail.SimpleMailMessage
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.function.Consumer

@Aspect
@Component
@Profile("email")
@Async
class NotificationAspect @Autowired constructor(private val emailService: EmailService<SimpleMailMessage>) {
    @After(
        "execution(* dev.lampirg.consultationappointment.service.student.AppointmentMaker.makeAppointment(..)) && " +
                "args(teacher,..,datePeriod)"
    )
    @Transactional(readOnly = true)
    fun notifyTeacherAboutAppointment(teacher: Teacher, datePeriod: DatePeriod) {
        if (datePeriod.appointments.isNotEmpty()) return
        val message = SimpleMailMessage()
        message.from = "consultation_service@pgups.example"
        message.subject = "ПГУПС: запись на консультацию"
        message.setTo(teacher.email)
        val text = """
                Здравствуйте, %s %s %s.
                На вашу консультацию, которая состоится %td.%tm с %tR до %tR записался студент.
                """.trimIndent()
        message.text = String.format(
            text,
            teacher.lastName, teacher.firstName, teacher.patronymic,
            datePeriod.startTime.toLocalDate(), datePeriod.startTime.toLocalDate(),
            datePeriod.startTime.toLocalTime(), datePeriod.endTime.toLocalTime()
        )
        emailService.sendEmail(message)
    }

    @After("execution(* dev.lampirg.consultationappointment.service.teacher.ConsultationMaker.deleteConsultation(..))")
    @Transactional(readOnly = true)
    fun notifyStudentsAboutDeletion(joinPoint: JoinPoint) {
        val datePeriod = joinPoint.args[1] as DatePeriod
        if (datePeriod.endTime.isBefore(LocalDateTime.now())) return
        val template = SimpleMailMessage()
        template.from = "consultation_service@pgups.example"
        template.subject = "ПГУПС: запись на консультацию"
        val text = """
                Здравствуйте, %s %s %s.
                К сожалению, консультация на которую вы записались (преподаватель: %s %s %s, дата: %td.%tm, время: с %tR до %tR), была отменена.
                """.trimIndent()
            datePeriod.appointments.forEach(Consumer<Appointment> { appointment: Appointment ->
                val student = appointment.student
                val teacher = appointment.teacher
                template.setTo(student.email)
                template.text = String.format(
                    text,
                    student.lastName, student.firstName, student.patronymic,
                    teacher.lastName, teacher.firstName, teacher.patronymic,
                    datePeriod.startTime.toLocalDate(), datePeriod.startTime.toLocalDate(),
                    datePeriod.startTime.toLocalTime(), datePeriod.endTime.toLocalTime()
                )
                emailService.sendEmail(template)
            })
    }
}